File: DemoMMOUnity/Assets/Game/Net/Script/Net/CsProtocol/Protocol/Equipment/EquipmentBaseData.cs
using System;
using ZJYFrameWork.Collection.Reference;

namespace ZJYFrameWork.Net.CsProtocol.Buffer.Protocol.Equipment
{
    public class EquipmentBaseData : Model, IPacket, IReference
    {
        /**
         * 介绍id
         */
        public int desId { get; set; }

        /**
         * 品阶
         */
        public int quality { get; set; }

        /**
         * 装备只能装配在什么位置
         */
        public int equipmentPosType { get; set; }

        /**
         * 圣遗物的名字
         */
        public string equipmentName { get; set; }

        /**
         * 主属性集合可以获取那些属性
         */
        public string mainAttributes { get; set; }


        public short ProtocolId()
        {
            return 209;
        }

        public void Clear()
        {
            desId = 0;
            quality = 0;
            equipmentPosType = -1;
            equipmentName = string.Empty;
            mainAttributes = string.Empty;
        }

        public static EquipmentBaseData ValueOf()
        {
            var data = ReferenceCache.Acquire<EquipmentBaseData>();
            data.Clear();
            return data;
        }

        public override void Unpack(byte[] bytes)
        {
            base.Unpack(bytes);  
        }
    }

    public class EquipmentBaseDataRegistration : IProtocolRegistration
    {
        public short ProtocolId()
        {
            return 209;
        }

        public void Write(ByteBuffer buffer, IPacket packet)
        {
            throw new NotImplementedException();
        }

        public IPacket Read(ByteBuffer buffer)
        {
            throw new NotImplementedException();
        }
    }
}

File: UnityMMOClient/Lua/application/app/net/webscoket/luaProtocol/euipment/equipmentBaseData.lua
---@class EquipmentBaseData
local EquipmentBaseData = class("EquipmentBaseData")
local this = EquipmentBaseData
function EquipmentBaseData:ctor()
    ---@type number
    self.desId = 0
    ---@type number
    self.quality = 0
    ---@type number
    self.equipmentPosType = 0
    ---@type string
    self.equipmentName = string.empty
    ---@type string
    self.mainAttributes = string.empty
end

---@param desId number 介绍id
---@param quality number 品阶
---@param equipmentPosType number 装备只能装配在什么位置
---@param equipmentName string 圣遗物的名字
---@param mainAttributes string 主属性集合可以获取那些属性
function EquipmentBaseData:new(desId, quality, equipmentPosType, equipmentName, mainAttributes)
    self.desId = desId
    self.quality = quality
    self.equipmentPosType = equipmentPosType
    self.equipmentName = equipmentName
    self.mainAttributes = mainAttributes
    return self
end

function EquipmentBaseData:protocolId()
    return 209
end

function EquipmentBaseData:write(buffer, packet)
    if packet == nil then
        return
    end
    local data = packet or EquipmentBaseData
    local message = {
        protocolId = data.protocolId(),
        packet = data
    }
    local jsonStr = JSON.encode(message)
    buffer:writeString(jsonStr)
end

function EquipmentBaseData:read(buffer)
    local jsonString = buffer:readString()
    ---字节读取器中存放字符
    ---@type {protocolId:number,packet:{desId:number, quality:number, equipmentPosType:number,equipmentName:string,mainAttributes:string}}
    local data = JSON.decode(jsonString)
    return EquipmentBaseData:new(
        data.packet.desId,
        data.packet.quality,
        data.packet.equipmentPosType,
        data.packet.equipmentName,
        data.packet.mainAttributes)
end

function EquipmentBaseData:readData(data)
    -- local jsonString = buffer:readString()
    -- ---字节读取器中存放字符
    -- ---@type {desId:number, quality:number, equipmentPosType:number,equipmentName:string,mainAttributes:string}
    -- local data = JSON.decode(jsonString)
    return EquipmentBaseData:new(
        data.desId,
        data.quality,
        data.equipmentPosType,
        data.equipmentName,
        data.mainAttributes)
end

return EquipmentBaseData
